### ISSUE TITLE
Engine support for headless Dart code

### ISSUE THREAD

**Issue by eseidelGoogle** [flutter/flutter#6192]
Several clients have asked for support of send-to-sync notifications, where a notification is received and the device is expected to sync down the new data in the background in response to the notification (instead of the notification itself carrying the payload).

My understanding from @collinjackson is that this is not currently possible to wire up due to lack of headless-Dart support in our APIs exposed to Flutter plugins.

This issue tracks work in the engine. Providing APIs to access the engine support is tracked in https://github.com/flutter/flutter/issues/3671

**Comment by tvolkert** [comment]
My understanding from talking to @collinjackson is that we need to decouple the concept of a Flutter application from its main view.

**Comment by collinjackson** [comment]
We can support this use case by modifying the Flutter engine to run Dart code outside the context of a view, and having the app be notified when a view is created. It will require changes at several layers of the platform and some modifications the Flutter API will be required.

**Comment by eseidelGoogle** [comment]
I spoke to @abarth just now and he mentioned that the "right" to add the ability to support headless code is to fix our window abstraction to not be a singleton.  When an isolate can draw to N windows, this is just the N=0 case.

**Comment by eseidelGoogle** [comment]
Note that it's possible to run headless code for send-to-sync notifications today in a Flutter-using app, it's just that that code would have to be Java or Objective-C and written twice.  This bug is about the engine work to make it possible to write headless sync code in Dart and only writing it once. :)

**Comment by sethladd** [comment]
cc @kasperl

**Comment by mit-mit** [comment]
@eseidel can you elaborate on what you mean by 'send-to-sync'? It's not entirely clear what is missing here.

**Comment by eseidelGoogle** [comment]
My understanding is it means running background code to "sync" data models on receipt of a notification w/o pulling up the app itself.  This is already possible in a flutter-using app, but requires writing Obj-C or Java code.  This but is about making it possible to run background/headless Dart code.

**Comment by mit-mit** [comment]
Got it. Dupe of https://github.com/flutter/flutter/issues/3671, or different from that ?

**Comment by mit-mit** [comment]
Closing this as a dupe of #3671

**Comment by a-siva** [comment]
Is the idea somewhat similar to what dartium had in terms of worker Dart isolates which could execute regular Dart code but did not have access to the DOM ?

**Comment by mehmetf** [comment]
This might be a somewhat obsolete request.

> tl;dr While send-to-sync notifications are still out there, they are being shunned in favor of complete notifications (i.e. they carry the entire payload necessary to render the notification). 

For instance, on iOS if your app is not running, iOS won't start it in order to receive the notification and sync data. Starting Android O, you have [5 seconds](https://developer.android.com/preview/features/background.html) to respond to an incoming message in the background and create a foreground service that shows notification.

**Comment by mit-mit** [comment]
I'm not entirely convinced this is obsolete. As you mention on Android O you can still run a background service, you just need to produce a notification. And iOS does allow tracking location in the background:
https://developer.apple.com/library/content/documentation/UserExperience/Conceptual/LocationAwarenessPG/CoreLocation/CoreLocation.html#//apple_ref/doc/uid/TP40009497-CH2-SW3

**Comment by mehmetf** [comment]
Originally the main reason this issue existed was send-to-sync notifications. The concept of send-to-sync is what I was calling obsolete. If this issue is being changed to be about running headless Dart code then https://github.com/flutter/flutter/issues/3671 adequately captures it. You can close this if you agree.

**Comment by mit-mit** [comment]
Yes, sorry I retitled it.

#3671 tracks nice abstractions for background execution, e.g. a platform independent way of saying "run this code every 30 mins". This tracks a small building block towards that: the ability to spawn Dart execution from platform code without needing a FlutterView.

Makes sense?

**Comment by mehmetf** [comment]
Sounds good.

**Comment by eseidelGoogle** [comment]
@cbracken mentioned to me in passing today that someone on @a-siva's team (might be @zanderso?) had started looking at this.  Is that true?

**Comment by sethladd** [comment]
@zanderso is assigned to https://github.com/flutter/flutter/issues/3671 I'm not 100% clear on the difference between this issue and 3671?

cc @mit-mit

**Comment by sethladd** [comment]
Oh, and I just heard from ClanHQ that they'd like this, too (or, they believe they need this) in order to help update some notification badges via Firebase Cloud Messaging + FlutterFire plugin, even if the app isn't in the foreground.

**Comment by mit-mit** [comment]
Sorry, I think my recent assignment of @zanderso to #3671 probably caused this confusion. I should have assigned the present bug instead.

The present bug tracks the most basic capability to spawn Dart execution from platform code without needing a FlutterView / Flutter app in the foreground. When this is done, you should be able to edit the android or iOS platform code to register a background process using their native mechanisms for that, and from that registration spawn the Dart execution so that you can write Dart code for your background code. @zanderso and others in the VM team are currently looking into what is required to support this.

#3671 tracks adding a more abstract Flutter API on top of that so that you can register background execution for common use cases without having to edit the platform-specific code, for example on a timer. We don't know yet exactly what these APIs should look like, and to what extent they should be cross-platform vs platform-specific. We also havn't committed to when we might work on #3671.

**Comment by zanderso** [comment]
I've landed a first cut of a plugin for the Android AlarmManager here: https://github.com/flutter/plugins/tree/master/packages/android_alarm_manager. Over the next week or so I'll be looking into extending its capabilities including exposing control over which Dart isolate alarm callbacks run on. Doing that will require further changes in the Flutter engine.

On the iOS side, we're still investigating. I've landed an API here to experiment with from plugins here: https://github.com/flutter/engine/commit/6c7350392306d9a654a82f3f048dd1c6f9b87e52, but we want to make the AlarmManager plugin a little more solid first before writing a plugin for iOS.

**Comment by zanderso** [comment]
Following https://github.com/flutter/engine/commit/7804e8588a031b6e85cf518b5fac07084661288f and https://github.com/flutter/plugins/commit/1cf9e9e7c9557947e1513cd7614291f74f20bb40. The Android AlarmManager plugin will now run callbacks on the same Isolate as an app's main activity. When the main activity is destroyed, the callbacks will continue to run on that Isolate. In my local testing, this Isolate persists across, e.g. hitting the back button, and turning off the screen. Further, when the main activity starts up again, we now check to see if a pre-existing Isolate should be used. For reference, see the plugin's [README.md](https://github.com/flutter/plugins/blob/master/packages/android_alarm_manager/README.md) and [example](https://github.com/flutter/plugins/tree/master/packages/android_alarm_manager/example) code.

The next step is to test various scenarios, in particular verifying/ensuring that Dart code running in callbacks in the absence of a main activity can still talk to plugins.

**Comment by sethladd** [comment]
Thanks @zanderso !

@pauldemarco please give it a try and let us know how it works for you. Thank you!

**Comment by zanderso** [comment]
After some testing, alarm callbacks successfully make calls to plugins (e.g. firebase_auth) when the Isolate was created by the main activity, but not when it was created by the background service. I'll be taking a look at making that work next.

**Comment by zanderso** [comment]
Following https://github.com/flutter/engine/commit/619f452ece9b8fec405c9b1c499f2509caaa79da, alarm callbacks can make calls to plugins when called from isolates created by a background service. It requires a little setup on the application side. See the alarm manager plugin example and the change here: https://github.com/flutter/plugins/commit/dd00f126447a539c5c796dd96639c303a8f84928.

The next step is to update first-party plugins to remove the assumption that there is always a foreground activity. I've started on that here: https://github.com/flutter/plugins/pull/299

**Comment by dragonfax** [comment]
Not having any luck getting this plugin to work. If I create a new project with flutter and follow the instructions in the README for the plugins, my alarm code never seems to fire. It tried both the version released on pub, and the latest version in the repo. It would help if someone would give it a quick try themselves as a sanity check for me.

**Comment by dragonfax** [comment]
Figured it out. The alarm function has to be a top-level function (i.e. named). Its even in the docs. I was using an anonymous function.

**Comment by mit-mit** [comment]
@zanderso can you please provide an update on the activity limitation comment from Dec 6th, and on the timeline for a solution for iOS?

**Comment by zanderso** [comment]
Where applicable, I've finished updating first-party plugins to remove the assumption that there is always a foreground activity. If you try to use a plugin that requires a foreground activity from a background service, an exception will be thrown.

I'll have to get back to you with a timeline for iOS side.

**Comment by cbracken** [comment]
@zanderso any updates on a timeline for the iOS side of this?

**Comment by Hixie** [comment]
This is currently waiting for Dart2 work to stabilize and for @chinmaygarde's engine embedding API refactor to land. We believe we are done with Android so far and the only thing that remains is iOS. We aren't sure we'll be able to do anything on iOS but we have an API we want to experiment with before we make a decision one way or the other.

**Comment by xster** [comment]
Should an example be part of the exit criteria? Also not sure whether https://github.com/flutter/flutter/issues/16490 should be duped in or if it's a separate step.

**Comment by MaikuB** [comment]
This sounds like a capability that I would need to wait to be supported. Been writing a notifications plugin but there were certain features I wasn't sure could be supported by the framework, particularly around notification actions. For example, a chat message is displayed as notification with the ability to send a direct reply (as per https://developer.android.com/training/notify-user/build-notification.html#reply-action). Correct me if i'm wrong but without the support for headless-Dart support then it's not possible to handle this scenario to invoke Dart code when the application has been terminated. Though looking at the history, it sounds like Android supports this.

**Comment by bkonyi** [comment]
It looks like this is a duplicate of #3671 which is complete.
